plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs

```diff
@@ -22,42 +21,50 @@ import           Data.Text.Utf16.Rope.Mixed       (Rope)
 import qualified Data.Text.Utf16.Rope.Mixed       as Rope
 import           Development.IDE.GHC.Compat
 import           Development.IDE.GHC.Error        (realSrcSpanToCodePointRange)
-import           Ide.Plugin.SemanticTokens.Types  (RangeIdSetMap)
+import           Ide.Plugin.SemanticTokens.Types  (HsSemanticTokenType (TModule),
+                                                   RangeHsSemanticTokenTypes (..))
 import           Language.LSP.Protocol.Types      (Position (Position),
                                                    Range (Range), UInt, mkRange)
 import           Language.LSP.VFS                 hiding (line)
 import           Prelude                          hiding (length, span)
 
 type Tokenizer m a = StateT PTokenState m a
+type HsSemanticLookup = Identifier -> Maybe HsSemanticTokenType
 
 
 data PTokenState = PTokenState
-  { rangeIdSetMap  :: !RangeIdSetMap,
-    rope           :: !Rope, -- the remains of rope we are working on
-    cursor         :: !Char.Position, -- the cursor position of the current rope to the start of the original file in code point position
-    columnsInUtf16 :: !UInt -- the column of the start of the current rope in utf16
+  {
+    rope                  :: !Rope -- the remains of rope we are working on
+    , cursor              :: !Char.Position -- the cursor position of the current rope to the start of the original file in code point position
+    , columnsInUtf16      :: !UInt -- the column of the start of the current rope in utf16
+    , rangeHsSemanticList :: [(Range, HsSemanticTokenType)] -- (range, token type) in reverse order
```

Comment:
can we move this out the state? It seems like all the other `Tokenizer` functions return `()`, instead we could just return `[(Range, HsSemanticTokenType)]`

Reply:
Sure, maybe we can do a cps style to collect them one by one.